### ISSUE TITLE
Unit-threaded and Dub fight each other

### ISSUE THREAD

**Issue by Abscissa** [atilaneves/unit-threaded#106]
I tried to use the unit-threaded test runner in a project by directly following the "Quick start with dub" instructions from the README, but it didn't work: Dub wouldn't compile the generated "bin/ut.d" file, and instead used its own test runner.

I think the problem is a combination of both dub and an outdated unit-threaded readme, so I'm pinging @s-ludwig (I hope that's the right person to ping, I know he's the original dub author, but I see a lot of commits there by other people now)

Here are the steps I went through:

- Upgraded dub from 1.6.0 to 1.7.1. Nothing changed.

- Added the line `sourceFiles "bin/ut.d"`. Nothing changed.

- Removed the line `mainSourceFile "bin/ut.d"`. This led to dub complaining that 
bin/ut.d wasn't found in any import path. Sounds like I'm starting to get somewhere, although by this point I'm already diverging from the instructions in unit-threaded's readme (that, presumably, used to work on some unknown version of dub).

- Added the line `importPaths "bin/"`. Now it's at least compiling ut.d, but dub is still trying to add its own test runner:

`bin/ut.d(5,5): Error: only one main allowed. Previously found main at /tmp/dub_test_root-00d69228-48f5-4f81-bdb3-a791e12823a9.d(26,12)`

- I don't see anything on code.dlang.org on disabling dub's test runner. So I make another shot-in-the-dark guess and change `dub test -c my-unit-threaded-config` to `dub run -c my-unit-threaded-config`. *Now* it seems to be working.

Oh, another fun wrinkle: I have a `void main() {}` in one of my source files (for when I compile unittests via rdmd). On my original non-unit-threaded test configuration, I have a `mainSourceFile` directive pointing at that file, so `dub test -c no-unit-threaded` ignores that file works fine. But if I add the same `mainSourceFile` directive to the same file on the "use unit-threaded" config, it *doesn't* exclude the file and complains about multiple main()...ugh, probably because I had to trick dub into thinking that config isn't a unittest...

(I'd also like to note, this has unfortunately been *very* characteristic of my experiences trying to use alternatives to *any* of dub's built-in features.)

**Comment by Abscissa** [comment]
>I tried to use the unit-threaded test runner in a project by directly following the "Quick start with dub" instructions from the README, 

Forgot to mention, the one maybe-notable change I made from the instructions was to update the requested version specifier for unit-threaded to the latest, v0.7.38

> Now it seems to be working.

Spoke too soon. The above changes got it to succesfully get me as far as unit-threaded's `--help` screen, but it didn't find any tests. Needed to also add `buildOptions "unittests"`.

**Comment by atilaneves** [comment]
I just copied the contents of the README and tried with both dub.json and dub.sdl and it worked as expected. Could you please post the exact contents of the dub.json/dub.sdl contents of your original attempt?

**Comment by Abscissa** [comment]
source/foobar/a.d:
```d
module foobar.a;
import std.stdio;

version(BLANK_MAIN) 
	void main() {}

unittest
{
	writeln("Hallo");
}
```

dub.sdl:
```
name "foobar"

// basic test runner
configuration "unittest" {
	targetType "executable"
	versions "BLANK_MAIN"
}

// unit-threaded test runner
configuration "unittest-unit-threaded" {
	targetType "executable"
	mainSourceFile "bin/ut.d"
	dependency "unit-threaded" version="~>0.7.38"
	preBuildCommands "dub run unit-threaded -c gen_ut_main -- -f bin/ut.d"
}
```

Cmdline:
```sh
$ dub test -c unittest-unit-threaded -- --help
Generating test runner configuration 'foobar-test-unittest-unit-threaded' for 'unittest-unit-threaded' (executable).
Excluding main source file bin/ut.d from test.
Performing "unittest" build using dmd for x86_64.
unit-threaded 0.7.38: target for configuration "library" is up to date.
foobar ~master: building configuration "foobar-test-unittest-unit-threaded"...
Running pre-build commands...
Building package unit-threaded in /home/nick/.dub/packages/unit-threaded-0.7.38/unit-threaded/
Performing "$DFLAGS" build using dmd for x86_64.
unit-threaded 0.7.38: target for configuration "gen_ut_main" is up to date.
To force a rebuild of up-to-date targets, run again with --force.
Running ../../../../.dub/packages/unit-threaded-0.7.38/unit-threaded/gen_ut_main -f bin/ut.d
Linking...
To force a rebuild of up-to-date targets, run again with --force.
Running ./foobar-test-unittest-unit-threaded --help
Hallo
All unit tests have been run successfully.
```

**Comment by atilaneves** [comment]
Huh. I looked into this some more and dub seems to completely ignore `mainSourceFile` for configurations other than `unittest`. There's nothing unit-threaded can do about that. I suggest swapping your configurations around - use unit-threaded in `unittest` and don't in the other.

**Comment by Abscissa** [comment]
Unfortunately, I can't use configuration "unittest" as the (reasonable) assumption is "dub test" will run a project's full test suite which, due to other dub limitations, doesn't work on my project. (My project is forced to use "dub test..." for only *some* of the tests and provide it's own script to run the full set of tests. Thus, I have a dummy "unittest" configuration that simply uses `preBuildCommands` to output a message directing the user to the correct command and generate a deliberate error to halt the tests.)

Luckily, however, the workaround I described above seems to work for me:

```
// Don't use "dub test", use:
//     dub run -c unittest-unit-threaded
configuration "unittest-unit-threaded" {
	targetType "executable"
	sourceFiles "bin/ut.d"
	importPaths "bin/"
	//mainSourceFile "bin/ut.d"
	buildOptions "unittests"
	dependency "unit-threaded" version="~>0.7.37"
	preBuildCommands "dub run unit-threaded -c gen_ut_main -- -f bin/ut.d"
}
```

**Comment by Abscissa** [comment]
>Huh. I looked into this some more and dub seems to completely ignore mainSourceFile for configurations other than unittest. There's nothing unit-threaded can do about that. I suggest swapping your configurations around - use unit-threaded in unittest and don't in the other.

Playing with this some more, I noticed that `gen_ut_main` runs `dub describe -c unittest` regardless of the configuration being used. Could that be why? Been awhile since I used this feature, but dub *should* be setting the configuration name into the DUB_CONFIG environment variable for all processes it launches ([see full list here](https://github.com/dlang/dub/blob/master/source/dub/generators/generator.d#L608)). `gen_ut_main` could read that and pass it back to `dub describe -c...`. But if it really is dub that's treating "unittest" differently, then I guess that wouldn't help this particular case.
